### PR TITLE
VR: fix issue between VPC VMs and other Public IPs in the same subnet as additional Public IPs

### LINKED ISSUE #8562: VPC Routing/Network Issue
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
VPC
~~~

##### CLOUDSTACK VERSION

~~~
4.18.1
~~~

##### CONFIGURATION

~~~
Advanced Zone, Network VPC 
~~~

##### OS / ENVIRONMENT

~~~
Cloudstack 4.18 on KVM
Multiple /27 IP's for Public 
~~~

##### SUMMARY

Unable to Reach  Public IP, which is in same deployment from VM's deployed in VPC

##### STEPS TO REPRODUCE

Cloudstack deployment is having multiple /27 IP addresses added to public Range. Deploy A VPC and acquire Multiple public IP from each /27 subnet. From any of the VM's deployed  in the VPC we are not able to reach the above public IP's except one.

checked the VPC router and the routing table is added as below
~~~
default via 99.127.xxx.65 dev eth1
10.20.1.0/24 dev eth5 proto kernel scope link src 10.20.1.1
10.20.2.0/24 dev eth6 proto kernel scope link src 10.20.2.1
169.254.0.0/16 dev eth0 proto kernel scope link src 169.254.198.170
99.127.xxx.32/27 dev eth2 proto kernel scope link src 99.127.xxx.51
99.127.xxx.64/27 dev eth1 proto kernel scope link src 99.127.xxx.93
99.127.xxx.96/27 dev eth3 proto kernel scope link src 99.127.xxx.125
~~~

From any VM's in VPC I can reach 99.127.xxx.64/27 subnet. Remaining public subnets are not reachable. From the VM We tried to ping the IP 99.127.xxx.107 (Firewall was Open to public for ICMP), but not pinging 

On capturing the packets we could see that packets are leaving the source network and its reaching VR with IP 99.127.xxx.107  but no response is received.

Packet Capture at source VPC router
~~~
03:34:16.950824 IP css1-cks-shared-1-node-18bb5b445d5 > 99.127.xxx.107: ICMP echo request, id 7168, seq 0, length 64
03:34:17.950957 IP css1-cks-shared-1-node-18bb5b445d5 > 99.127.xxx.107: ICMP echo request, id 7168, seq 1, length 64
~~~
Received packets at 99.127.xxx.107 VR
~~~
03:34:16.950549 IP 10.20.2.137 > 99.127.xxx.107: ICMP echo request, id 7168, seq 0, length 64
03:34:16.951536 IP 99.127.xxx.107 > 10.20.2.137: ICMP echo reply, id 7168, seq 0, length 64
03:34:17.950598 IP 10.20.2.137 > 99.127.xxx.107: ICMP echo request, id 7168, seq 1, length 64
03:34:17.951405 IP 99.127.xxx.107 > 10.20.2.137: ICMP echo reply, id 7168, seq 1, length 64
~~~
On above we could see that the source IP address is marked as '10.20.2.137' and from the 99.127.xxx.107 router this Private IP will not be reachable.  Ideally It should be the Public IP address of VPC.

Please let us know your comments on this.

### PR DESCRIPTION
**Pull request by weizhouapache**
### Description

This PR fixes #8562 

If there are mutliple public subnets in a VR, there should be iptables rules like
```
-A POSTROUTING -o eth1 -j SNAT --to-source <Source NAT IP>
-A POSTROUTING -o eth3 -j SNAT --to-source <Source NAT IP>
-A POSTROUTING -o ethX -j SNAT --to-source <Source NAT IP>
```

in VR of isolated networks, these rules do exist.
Howver in VPC VR, only the first rule exists, the other rules are missing

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test rocky8 kvm-rocky8

File: systemvm/debian/opt/cloud/bin/cs/CsAddress.py [hunk 1]
```diff
@@ -453,8 +453,8 @@ def fw_router(self):
                 ["", "", "-A NETWORK_STATS_%s -o %s ! -i eth0 -p tcp" % (self.dev, self.dev)])
             self.fw.append(
                 ["", "", "-A NETWORK_STATS_%s -i %s ! -o eth0 -p tcp" % (self.dev, self.dev)])
-            self.fw.append(["nat", "",
-                            "-A POSTROUTING -o %s -j SNAT --to-source %s" % (self.dev, self.cl.get_eth2_ip())])
+            self.fw.append(
+                ["nat", "", "-A POSTROUTING -o %s -j SNAT --to-source %s" % (self.dev, self.cl.get_eth2_ip())])
```
Comment (DaanHoogland) [inline systemvm/debian/opt/cloud/bin/cs/CsAddress.py:457]: is this an intended change?

File: server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java [hunk 2]
```diff
@@ -2146,6 +2147,12 @@ public boolean finalizeVirtualMachineProfile(final VirtualMachineProfile profile
                 " on the virtual router.", RouterLogrotateFrequency.key(), routerLogrotateFrequency, dc.getUuid()));
         buf.append(String.format(" logrotatefrequency=%s", routerLogrotateFrequency));
 
+        if (router.getVpcId() != null) {
+            List<IPAddressVO> vpcIps = _ipAddressDao.listByAssociatedVpc(router.getVpcId(), true);
+            if (CollectionUtils.isNotEmpty(vpcIps)) {
+                buf.append(String.format(" source_nat_ip=%s", vpcIps.get(0).getAddress().toString()));
+            }
+        }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java:2155]: this bit should be in the method `VpcVirtualNetworkApplianceManagerImpl.finalizeVirtualMachineProfile(..)`.

best in a call to this.appendSourceNats()

File: systemvm/debian/opt/cloud/bin/cs/CsAddress.py (hunk 1 above)
Comment (weizhouapache) [reply systemvm/debian/opt/cloud/bin/cs/CsAddress.py:457]: Just a reformat, nothing changed actually.
Can be reverted

File: server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java (hunk 2 above)
Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java:2155]: Good point, will have a look

**Review by sureshanaparti — APPROVED** [review]
code lgtm

File: systemvm/debian/opt/cloud/bin/cs/CsAddress.py (hunk 1 above)
Comment (DaanHoogland) [reply systemvm/debian/opt/cloud/bin/cs/CsAddress.py:457]: ok, never mind.

File: server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java [hunk 3]
```diff
@@ -294,7 +295,23 @@ public boolean finalizeVirtualMachineProfile(final VirtualMachineProfile profile
             }
         }
 
-        return super.finalizeVirtualMachineProfile(profile, dest, context);
+        super.finalizeVirtualMachineProfile(profile, dest, context);
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:298]: just checking, we can safely ignore the return value here?

Comment (weizhouapache) [reply server/src/main/java/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java:298]: yes @DaanHoogland 

it always return true in the parent class, 

https://github.com/apache/cloudstack/blob/37bb24f29e02480488312f1a0c8965c3543651d8/server/src/main/java/com/cloud/network/router/VirtualNetworkApplianceManagerImpl.java#L2153

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@blueorangutan test

**Pull request merged**